File: Diagrammatize.Design.Pattern/code/cxx/Proxy/Sample/Printer.cpp
#include "Printer.hpp"

#include <iostream>
#include <chrono>
#include <thread>

Printer::Printer() {
    heavyJob("正在生成Printer的实例");
}

Printer::Printer(const std::string &name_): name(name_) {
    heavyJob("正在生成Printer的实例(" + name + ")");
}

void Printer::setPrinterName(const std::string& name_) {
    name = name_;
}

std::string Printer::getPrinterName() {
    return name;
}

void Printer::print(const std::string& string) {
    std::cout << "=== " << name << " ===" << std::endl;
    std::cout << string << std::endl;
}

void Printer::heavyJob(const std::string& msg) {
    std::cout << msg << std::flush;
    for (int i = 0; i < 5; i++) {
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
        std::cout << "." << std::flush;
    }
    std::cout << "结束。" << std::endl;
}

File: Diagrammatize.Design.Pattern/code/python/Proxy/Sample/Printer.py
import time

from Printable import Printable

class Printer(Printable):
    def __init__(self, name=""):
        self.name = name
        if self.name:
            self.heavyJob("正在生成Printer的实例(" + name + ")")
        else:
            self.heavyJob("正在生成Printer的实例")

    def setPrinterName(self, name):
        self.name = name

    def getPrinterName(self):
        return self.name

    def print(self, string):
        print("=== " + self.name + " ===")
        print(string)

    def heavyJob(self, msg):
        print(msg, end="", flush=True)
        for i in range(5):
            time.sleep(1.0)
            print(".", end="", flush=True)
        print("结束。")
